### PR TITLE
fix(dataset): Use platform delimiter to compose the browse paths

### LINKED ISSUE #1956: Use PlatformInfo to better compose browse paths
**Is your feature request related to a problem? Please describe.**
Dataset names that have `.` in them are split by the browse endpoint regardless of the [Platform delimiter](https://github.com/linkedin/datahub/blob/f0485a490e08445d1459de3cd697c6cc0b2189d5/metadata-models/src/main/pegasus/com/linkedin/dataplatform/DataPlatformInfo.pdl#L72). See the bread crumbs in the image below:

image

I didn't expect to see the last two crumbs to be `a` and `test`. Instead I expected to see one unified crumb `a.test`. 

**Describe the solution you'd like**
It seems like we always split on `.` [here](https://github.com/linkedin/datahub/blob/master/metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java#L32) but perhaps we can use the the [Platform delimiter](https://github.com/linkedin/datahub/blob/f0485a490e08445d1459de3cd697c6cc0b2189d5/metadata-models/src/main/pegasus/com/linkedin/dataplatform/DataPlatformInfo.pdl#L72) to be smarter about this?

The desired end result would be to treat `.` in dataset names as any other alphanumeric character for platforms that use a different delimiter.

**Describe alternatives you've considered**
No alternatives have been considered

**Additional context**
Using `.` in names is not very uncommon. Supporting them properly would facilitate DataHub adoption.

### PR DESCRIPTION
**Pull request by nagarjunakanamarlapudi**
fix(dataset): Use platform delimiter to compose the browse paths. Fixes #1956 

RestliClient is passed to the dataset index builder.  Hence it can call the /datasetPlatform API to get the platform delimiter when constructing browse paths. 

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java [hunk 1]
```diff
@@ -1,35 +1,84 @@
...
+    final String dataset = urn.getDatasetNameEntity();
+    String browsePath = "/" + dataOrigin + "/" + platform + "/" + dataset;
+    final String delimiter = DELIMITER_CACHE.get(urn.getPlatformEntity());
+    log.debug("delimiter for platform {}, is {} ", urn.getPlatformEntity().getPlatformNameEntity(), delimiter);
+    if (!delimiter.isEmpty()) {
+      final Character delimiterChar = delimiter.charAt(0);
+      if (delimiterChar.equals('/')) {
+        browsePath = "/" + dataOrigin + "/" + platform + dataset;
+      } else {
+        browsePath = ("/" + dataOrigin + "/" + platform + "/" + dataset).replace(delimiterChar, '/');
```
Comment (jplaisted) [inline metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java:78]: Shouldn't this be `... + dataset.replace(delimiterChar, '/')`, not the entire thing?

File: metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java [hunk 2]
```diff
@@ -1,35 +1,84 @@
...
+    final String dataOrigin = urn.getOriginEntity().name();
+    final String platform = urn.getPlatformEntity().getPlatformNameEntity();
+    final String dataset = urn.getDatasetNameEntity();
+    String browsePath = "/" + dataOrigin + "/" + platform + "/" + dataset;
+    final String delimiter = DELIMITER_CACHE.get(urn.getPlatformEntity());
+    log.debug("delimiter for platform {}, is {} ", urn.getPlatformEntity().getPlatformNameEntity(), delimiter);
+    if (!delimiter.isEmpty()) {
+      final Character delimiterChar = delimiter.charAt(0);
+      if (delimiterChar.equals('/')) {
+        browsePath = "/" + dataOrigin + "/" + platform + dataset;
```
Comment (jplaisted) [inline metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java:76]: This won't change browsePath, so just remove this condition.

File: metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java [hunk 3]
```diff
@@ -1,35 +1,84 @@
...
 import javax.annotation.Nonnull;
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
 
 @Slf4j
 public class DatasetIndexBuilder extends BaseIndexBuilder<DatasetDocument> {
-  public DatasetIndexBuilder() {
+  private static RestliRemoteDAO<DataPlatformSnapshot, DataPlatformAspect, DataPlatformUrn> dataPlatformDAO;
```
Comment (jplaisted) [inline metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java:38]: Why is this static?

File: metadata-builders/src/test/java/com/linkedin/metadata/builders/search/DatasetIndexBuilderTest.java [hunk 4]
```diff
@@ -14,67 +14,119 @@
...
+        .thenReturn(Optional.of(
+            new DataPlatformInfo().setName("foo").setType(PlatformType.FILE_SYSTEM).setDatasetNameDelimiter("/")));
+
+    Mockito.when(dataPlatformDAO.get(Mockito.eq(DataPlatformInfo.class), Mockito.eq(new DataPlatformUrn("hive"))))
+        .thenReturn(Optional.of(
+            new DataPlatformInfo().setName("hive").setType(PlatformType.FILE_SYSTEM).setDatasetNameDelimiter(".")));
+
+    indexBuilder = new DatasetIndexBuilder(mockRestClient);
+
+    FieldSetter.setField(indexBuilder, indexBuilder.getClass().getDeclaredField("dataPlatformDAO"), dataPlatformDAO);
```
Comment (jplaisted) [inline metadata-builders/src/test/java/com/linkedin/metadata/builders/search/DatasetIndexBuilderTest.java:79]: All this mocking and setting of private fields indicates your API here is probably in need of a refactor. These are brittle practices and antipatterns.

- If the field name is ever changed or the field removed, or made final, tests will fail rather than a compile time error. _Private fields are not part of an API and should not be access via reflection._ Generally speaking _avoid reflection_. There's usually a more elegant way of expressing what you want without it, and in a less brittle way too.
- go/lessmocks https://testing.googleblog.com/2013/05/testing-on-toilet-dont-overuse-mocks.html


Suggestions:

- Create a small interface that, given a data platform urn, returns the delimiter.
- Create one implementation of that interface that uses restli to get it. Move the cache here too.
- In tests you can use a fake implementation, even just a lambda...
- Have `DatasetIndexBuilder` take the interface in its constructor. You can even have multiple constructors if you want, one that takes this interface (for tests) and one that takes a RestClient and makes the rest implementation above (for production).

No more mocks or awkward setting of private fields. Also means private fields can be `final`.

tl;dr

```
interface DatasetPlatformPaths {
  String getDelimiter(DataPlatformUrn urn);
}
```

```
final class RestliDatasetPlatformPaths implements DatasetPlatformPaths {
  // implementation goes here, include the cache here, etc
  RestliDatasetPlatformPaths(Client client) {...}
}
```

```
class DatasetIndexBuilder {
  // Visible for testing
  DatasetIndexBuilder(@Nonnull DatasetPlatformPaths restliClient) {
    ...
  }
  public DatasetIndexBuilder(@Nonnull Client restliClient) {
    this(new RestliDatasetPlatformPaths(client));
  }
}
```

And now tests become:

```
indexBuilder = new DatasetIndexBuilder((urn) => {
  // ... return "." if hive, "/" if foo, etc
});
```

**Review by jplaisted — CHANGES_REQUESTED** [review]

File: metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java (hunk 1 above)
Comment (jplaisted) [reply metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java:78]: So might be worth adding a test here too.

If my platform is something like `aws:s3`, delimiter `:`, and dataset `foo:bar`, what would this return? I'd assume `/origin/aws:s3/foo/bar` is expected, but this returns `/origin/aws/s3/foo/bar`.

File: metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java (hunk 2 above)
Comment (jplaisted) [reply metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java:76]: Ah, nevermind, it accounts for the leading `/`.

Which brings up a good point. What happens to other delimiters that also have a leading delimiter? e.g. `:foo:bar`? Wouldn't you end up with `/origin/platform//foo/bar`?

**Review by jplaisted — CHANGES_REQUESTED** [review]

**Comment by jjoyce0510** [comment]
@nagarjunakanamarlapudi Is this still under active development? If not, can we close this PR for cleanliness? 

cc. @shirshanka

**Comment by jjoyce0510** [comment]
Ping on this one. Seeing that https://github.com/linkedin/datahub/issues/1956 depends on this

**Comment by nagarjunakanamarlapudi** [comment]
I am seeing rebase issues. I will close this PR and create a new one
